File: Pharmaceutical_Company_Example/03_Behavioral/R/Observer.R
# Observer Interface
Observer <- setRefClass(
  "Observer",
  methods = list(
    update = function(state) {
      stop("This method should be overridden by concrete observers!")
    }
  )
)

# Subject
DrugApprovalProcess <- setRefClass(
  "DrugApprovalProcess",
  fields = list(observers = "list", state = "character"),
  methods = list(
    attach_observer = function(observer) {
      observers[[length(observers) + 1]] <<- observer
    },
    detach_observer = function(observer) {
      observers <<- observers[!sapply(observers, identical, observer)]
    },
    notify_observers = function() {
      message("DrugApprovalProcess: Notifying observers about state change to:", state)
      for (observer in observers) {
        observer$update(state)
      }
    },
    set_state = function(new_state) {
      state <<- new_state
      message("DrugApprovalProcess: State changed to:", state)
      notify_observers()
    }
  )
)

# Clinical Trials Department
ClinicalTrials <- setRefClass(
  "ClinicalTrials",
  contains = "Observer",
  methods = list(
    update = function(state) {
      message("Clinical Trials Department: Notified about state change to:", state)
    }
  )
)

# Quality Assurance Department
QualityAssurance <- setRefClass(
  "QualityAssurance",
  contains = "Observer",
  methods = list(
    update = function(state) {
      message("Quality Assurance Department: Notified about state change to:", state)
    }
  )
)

# Legal Department
LegalDepartment <- setRefClass(
  "LegalDepartment",
  contains = "Observer",
  methods = list(
    update = function(state) {
      message("Legal Department: Notified about state change to:", state)
    }
  )
)

# Regulatory Affairs Department
RegulatoryAffairs <- setRefClass(
  "RegulatoryAffairs",
  contains = "Observer",
  methods = list(
    update = function(state) {
      message("Regulatory Affairs Department: Notified about state change to:", state)
    }
  )
)

drug_approval_process <- DrugApprovalProcess$new()

# Create observers
clinical_trials <- ClinicalTrials$new()
quality_assurance <- QualityAssurance$new()
legal <- LegalDepartment$new()
regulatory_affairs <- RegulatoryAffairs$new()

# Attach observers to the subject
drug_approval_process$attach_observer(clinical_trials)
drug_approval_process$attach_observer(quality_assurance)
drug_approval_process$attach_observer(legal)
drug_approval_process$attach_observer(regulatory_affairs)

# Simulate state changes and notify observers
drug_approval_process$set_state("Clinical Trial Started")
drug_approval_process$set_state("Clinical Trial Completed")
drug_approval_process$set_state("Quality Check in Progress")
drug_approval_process$set_state("Under Regulatory Review")
drug_approval_process$set_state("Approved")

File: Pharmaceutical_Company_Example/03_Behavioral/JavaScript/Observer.js
// Observer Interface
class Observer {
  update(state) {
    throw new Error("This method should be overridden by concrete observers!");
  }
}

// Subject
class DrugApprovalProcess {
  constructor() {
    this._observers = [];
    this._state = "";
  }

  attachObserver(observer) {
    this._observers.push(observer);
  }

  detachObserver(observer) {
    this._observers = this._observers.filter(obs => obs !== observer);
  }

  notifyObservers() {
    console.log(`DrugApprovalProcess: Notifying observers about state change to: ${this._state}`);
    for (const observer of this._observers) {
      observer.update(this._state);
    }
  }

  setState(newState) {
    this._state = newState;
    console.log(`DrugApprovalProcess: State changed to: ${this._state}`);
    this.notifyObservers();
  }
}

// Concrete Observer: Clinical Trials Department
class ClinicalTrials extends Observer {
  update(state) {
    console.log(`Clinical Trials Department: Notified about state change to: ${state}`);
  }
}

// Concrete Observer: Quality Assurance Department
class QualityAssurance extends Observer {
  update(state) {
    console.log(`Quality Assurance Department: Notified about state change to: ${state}`);
  }
}

// Concrete Observer: Legal Department
class LegalDepartment extends Observer {
  update(state) {
    console.log(`Legal Department: Notified about state change to: ${state}`);
  }
}

// Concrete Observer: Regulatory Affairs Department
class RegulatoryAffairs extends Observer {
  update(state) {
    console.log(`Regulatory Affairs Department: Notified about state change to: ${state}`);
  }
}

// Create the DrugApprovalProcess (Subject)
const drugApprovalProcess = new DrugApprovalProcess();

// Create observers
const clinicalTrials = new ClinicalTrials();
const qualityAssurance = new QualityAssurance();
const legal = new LegalDepartment();
const regulatoryAffairs = new RegulatoryAffairs();

// Attach observers to the subject
drugApprovalProcess.attachObserver(clinicalTrials);
drugApprovalProcess.attachObserver(qualityAssurance);
drugApprovalProcess.attachObserver(legal);
drugApprovalProcess.attachObserver(regulatoryAffairs);

// Simulate state changes and notify observers
drugApprovalProcess.setState("Clinical Trial Started");
drugApprovalProcess.setState("Clinical Trial Completed");
drugApprovalProcess.setState("Quality Check in Progress");
drugApprovalProcess.setState("Under Regulatory Review");
drugApprovalProcess.setState("Approved");
